runtime/ts/storage/pouchdb/pouch-db-variable.ts

```diff
@@ -0,0 +1,340 @@
+      
+      // compare revisions
+      if (this._rev !== result._rev) {
+        // remote revision is different, update local copy.
+        this._stored = result['value'];
+        this._rev = result._rev;
+        this.version++;  // yuck.
+        // TODO(lindner): fire change events here?
+      }
+    } catch (err) {
+      if (err.name === 'not_found') {
+        this._stored = null;
+        this._rev = undefined;
+      } else {
+        console.warn('PouchDbVariable.getStored err=', err);
+        throw err;
+      }
+    }
+    return this._stored;
+  }
+
+  /**
+   * Provides a way to apply changes to the stored value in a way that
+   * will result in the stored value being written to the underlying
+   * PouchDB.
+   * 
+   * - A new entry is stored if it doesn't exists.
+   * - If the existing entry is available it is fetched
+   * - If revisions differ a new item is written.
+   * - The storage is potentially mutated and written.
+   *
+   * @param variableStorageMutator allows for changing the variable.
+   * @return the current value of _stored.
+   */
+  private async getStoredAndUpdate(variableStorageMutator: VariableStorageMutator): Promise<VariableStorage> {
+    // Keep retrying the operation until it succeeds.
+    while (1) {
+      let doc;
+
+      let notFound = false;
+      try {
+        doc = await this.db.get(this.pouchDbKey.location);
+        // Check remote doc.
+        // TODO(lindner): refactor with getStored above.
+        if (this._rev !== doc._rev) {
+          // remote revision is different, update local copy.
+          this._stored = doc['value'];
+          this._rev = doc._rev;
+          this.version++;  // yuck.
+          // TODO(lindner): fire change events here?
+        }
+      } catch (err) {
+        if (err.name !== 'not_found') {
+          throw err;
+        }
+        notFound = true;
+        // setup basic doc, model/version updated below.
+        doc = {
+          _id: this.pouchDbKey.location
+        };
+      }
+
+      // Run the mutator on a copy of the existing model
+      const newValue = variableStorageMutator({ ...this._stored });
+      
+      // Check if the mutator made any changes..
+      // TODO(lindner): add a deep equals method for VariableStorage
+      if (!notFound && JSON.stringify(this._stored) === JSON.stringify(newValue)) {
+        // mutator didn't make any changes.
+        return this._stored;
+      }
+      
+      // Apply changes made by the mutator
+      doc['value'] = newValue;
+      doc['version'] = this.version;
+
+      // Update on pouchdb
+      try {
+        const putResult = await this.db.put(doc);
+        console.log('VAR PUTRESULT', putResult, doc);
+        // success! update local with new stored value
+        this._rev = putResult.rev;
+        this._stored = newValue;
+        
+        return this._stored;
+      } catch (err) {
+        if (err.name === 'conflict') {
+          // keep trying;
```

Comment:
Similar to collection, should we bail out after some number of attempts or at least log?

Reply:
'conflict' is fine, it just means that someone wrote before you.  Solution is to refetch and reapply your change, which is not a problem because we use CRDTs.

Added TODO to the the top of the loop to add backoff/give up after set period of time.